std/range/package.d

```diff
@@ -1574,18 +1574,20 @@ Returns:
     alias of that range's type.
  */
 auto chooseAmong(Ranges...)(size_t index, Ranges rs)
-if (Ranges.length > 2
-        && is(typeof(choose(true, rs[0], rs[1])))
-        && is(typeof(chooseAmong(0, rs[1 .. $]))))
+if (Ranges.length >= 2 && allSatisfy!(isInputRange, staticMap!(Unqual, Ranges)))
 {
-    return choose(index == 0, rs[0], chooseAmong(index - 1, rs[1 .. $]));
-}
-
-/// ditto
-auto chooseAmong(Ranges...)(size_t index, Ranges rs)
-if (Ranges.length == 2 && is(typeof(choose(true, rs[0], rs[1]))))
-{
-    return choose(index == 0, rs[0], rs[1]);
+    static if (Ranges.length == 2 && is(typeof(choose(true, rs[0], rs[1]))))
+    {
+        return choose(index == 0, rs[0], rs[1]);
+    }
+    else static if (is(typeof(choose(true, rs[0], rs[1]))) && is(typeof(chooseAmong(0, rs[1 .. $]))))
+    {
+        return choose(index == 0, rs[0], chooseAmong(index - 1, rs[1 .. $]));
+    }
+    else
+    {
+        static assert(0, "The ranges have no common type.");
```

Comment:
Hehe this paradigm switch has happened a couple of months ago, see e.g. @quickfur's excellent explananation at https://github.com/dlang/phobos/pull/5148#issuecomment-286839494:

> And just for the record, the last bit is what I meant in various forum posts that user-facing overloads should have sig constraints that reflect *intent* rather than implementation: the intent of `each` is that it would accept anything that either supports `foreach` or has a range API.  So those should be what constitute the sig constraints of `each`. 
> There may be, however, some ranges that satisfy these constraints but which aren't handled by the current implementation; these should be caught by the final else clause in the static if inside each overload body, with a helpful error message explaining to the user why it didn't work ("support for range with characteristics XYZ isn't implemented yet", or "cannot call each with range with characteristics XYZ because PQR").

In this case one could argue for both sides, but imho having a concrete message with what went wrong helps the user a lot more than "X didn't match constraints".
I don't feel strongely about this particular case though, so I'm happy to change it as suggested if that's the only thing that blocks this PR. Other Opinions?